MIP/mip-58/README.md

```diff
@@ -0,0 +1,462 @@
+        bytes32[] memory bridgeTransferIds,
+        bytes32[] memory initiators,
+        address[] memory recipients,
+        uint256[] memory amounts,
+        uint256[] memory nonces
+    ) external onlyRole(RELAYER_ROLE) {
+        uint256 length = bridgeTransferIds.length;
+        // checks if all arrays are of the same length
+        require(
+            initiators.length == length && recipients.length == length && amounts.length == length
+                && nonces.length == length,
+            InvalidLenghts()
+        );
+        // iterate over the arrays and complete the bridge transfer
+        for (uint256 i; i < length; i++) {
+            _completeBridgeTransfer(bridgeTransferIds[i], initiators[i], recipients[i], amounts[i], nonces[i]);
+        }
+    }
+
+    function _completeBridgeTransfer(
+        bytes32 bridgeTransferId,
+        bytes32 initiator,
+        address recipient,
+        uint256 amount,
+        uint256 nonce
+    ) internal {
+        // _l2l1RateLimit(amount);
+        // Ensure the bridge transfer ID is valid against the initiator, recipient, amount, and nonce
+        require(
+            bridgeTransferId == keccak256(abi.encodePacked(initiator, recipient, amount, nonce)),
+            InvalidBridgeTransferId()
+        );
+        // Ensure the bridge transfer has not already been completed
+        require(noncesToIncomingBridgeTransferIds[nonce] == bytes32(0x0), CompletedBridgeTransferId());
+
+        // Store the nonce to bridge transfer ID
+        noncesToIncomingBridgeTransferIds[nonce] = bridgeTransferId;
+
+        // Transfer the MOVE tokens to the recipient
+        if (!moveToken.transfer(recipient, amount)) revert MOVETransferFailed();
+
+        emit BridgeTransferCompleted(bridgeTransferId, initiator, recipient, amount, nonce);
+    }
+
+    function togglePause() external onlyRole(DEFAULT_ADMIN_ROLE) {
+        paused() ? _pause() : _unpause();
+    }
+}
+
+```
+
+```move
+// Based on https://github.com/movementlabsxyz/aptos-core/blob/movement/aptos-move/framework/aptos-framework/sources/atomic_bridge.move  
+
+    /// Module-level resource to store the nonce  
+    struct Nonce has key {  
+        value: u64,  
+    }  
+
+    /// Creates bridge transfer details with validation.  
+    ///  
+    /// @param initiator The initiating party of the transfer.  
+    /// @param recipient The receiving party of the transfer.  
+    /// @param amount The amount to be transferred.  
+    /// @param nonce The unique nonce for the transfer.  
+    /// @return A `BridgeTransferDetails` object.  
+    /// @abort If the amount is zero.  
+    internal fun create_details<Initiator: store, Recipient: store>(  
+        initiator: Initiator,  
+        recipient: Recipient,  
+        amount: u64,  
+        nonce: u64  
+    ): BridgeTransferDetails<Initiator, Recipient> {  
+        assert!(amount > 0, EZERO_AMOUNT);  
+
+        BridgeTransferDetails {  
```

Comment:
there is not transfer status in the details.
How do we prevent someone relaying a relayer Tx with the same parameters?

Reply:
we can't have an status because transaction happens on source then on target. What we do is we check if a value has been placed on that bridgeTransferId, if it has, then it has already been done.

Reply:
I don't see how this is implemented in the code.

Reply:
```solidity
 require(nonce > 0, InvalidNonce());
        require(idsToIncomingNonces[bridgeTransferId] == 0, CompletedBridgeTransferId());
        // Ensure the bridge transfer ID is valid against the initiator, recipient, amount, and nonce
        require(
            bridgeTransferId == keccak256(abi.encodePacked(initiator, recipient, amount, nonce)),
            InvalidBridgeTransferId()
        );

        // Store the nonce to bridge transfer ID
        idsToIncomingNonces[bridgeTransferId] = nonce;
 ```
 
 I feel like solidity is easier to understand, so the move side should have the same. The only thing missing is the assertion that nonce is over 0